[CollapsingToolbarLayout] Allow multiple lines of text in expanded state

johan12345 opened the pull request:
closes #6 and was also discussed in https://issuetracker.google.com/issues/136120586, where you said you would welcome a PR.

We have implemented the possibility to use multiple lines of text as a title in the collapsing app bar. The additional lines are displayed in the expanded state and will be hidden in the collapsed state, like this:

![Demo image](https://raw.githubusercontent.com/opacapp/multiline-collapsingtoolbar/master/demo.gif)

This is based on the work of @raphaelm and me as well as a few [contributors](https://github.com/opacapp/multiline-collapsingtoolbar/graphs/contributors) in the [multiline-collapsingtoolbar](https://github.com/opacapp/multiline-collapsingtoolbar) library, which is based on the CollapsingToolbarLayout code of the Android Design Support library (predecessor of Material Components) and has already become quite popular.

However, as discussed in https://github.com/opacapp/multiline-collapsingtoolbar/issues/62, it would be quite difficult to update our library with support for the new Material Components and AndroidX libraries, as many of the interfaces we are calling are not part of the public API and thus can only be called from the Material Components library itself. Therefore, it would be great if this feature could instead be included in the official Material Components library.

In addition to copying over the adjustments from our library into the CollapsingTextHelper codebase, I also added a demo in the catalog app to showcase this new feature.

johan12345 commented:
As the Travis CI build is also failing on `master`, I'm assuming these test failures are not due to my changes.

cketcham commented on lib/java/com/google/android/material/internal/CollapsingTextHelper.java:
```diff
@@ -691,13 +783,56 @@ private void calculateUsingTextSize(final float textSize) {
+        truncatedText = TextUtils.concat(textBefore, truncatedLineText);
+
+      } else {
+        truncatedText = text;
```
In this case should it potentially be ellipsized?

johan12345 replied:
Umm, the `else` case is when the number of lines required to show the whole title (`layout.getLineCount()`) is `<= maxLines` - so the whole text can be shown. Why should it be ellipsized then?

cketcham replied:
yeah think I turned that logic around in my head! Looks right.

cketcham commented on lib/java/com/google/android/material/internal/CollapsingTextHelper.java:
```diff
@@ -571,34 +628,62 @@ public void draw(Canvas canvas) {
       final boolean drawTexture = useTexture && expandedTitleTexture != null;
```
If we care enough to check that expandedTitleTexture exists, maybe we should also check that the other textures exist?

johan12345 replied:
Yes, that makes sense. Fixed in 5c073ae. 🙂

cketcham commented on lib/java/com/google/android/material/internal/CollapsingTextHelper.java:
```diff
@@ -668,15 +768,8 @@ private void calculateUsingTextSize(final float textSize) {
       // collapsed text size
       final float scaledDownWidth = expandedWidth * textSizeRatio;
```
I think this variable is no longer read?

johan12345 replied:
Oh, right. I removed it in a0cca8a.

cketcham commented:
@johan12345 When I run the demo in the catalog I see the text is black instead of white. Do you see that as well?

cketcham commented:
It looks as though there might be a bug in RTL as well? If you enable [pseudolocales](https://developer.android.com/guide/topics/resources/pseudolocales) for the catalog, it causes overlapping text: 
![Screen Shot 2019-07-10 at 4 05 29 PM](https://user-images.githubusercontent.com/16722/61000958-8f534b80-a32c-11e9-957e-ba2ef355b78f.png)
![Screen Shot 2019-07-10 at 4 05 16 PM](https://user-images.githubusercontent.com/16722/61000959-8febe200-a32c-11e9-89a4-ead5f00e526c.png)
![Screen Shot 2019-07-10 at 4 05 00 PM](https://user-images.githubusercontent.com/16722/61000960-8febe200-a32c-11e9-8a29-6eb59a7c1cc3.png)

johan12345 commented:
> @johan12345 When I run the demo in the catalog I see the text is black instead of white. Do you see that as well?

Yes, I see that as well, but the same happens with the original Collapsing Demo, also on `master`. Seems like the default style for the CollapsingToolbarLayout is not correctly set. It works correctly when switching to the dark theme in the main activity.

dsn5ft commented:
> > @johan12345 When I run the demo in the catalog I see the text is black instead of white. Do you see that as well?
> 
> Yes, I see that as well, but the same happens with the original Collapsing Demo, also on `master`. Seems like the default style for the CollapsingToolbarLayout is not correctly set. It works correctly when switching to the dark theme in the main activity.

@johan12345 Ah you can set a theme overlay on your AppBarLayout (`android:theme="@style/ThemeOverlay.Catalog.AppBarLayout"`) to fix the text color issue, similar to [cat_topappbar_collapsing_fragment.xml](https://github.com/material-components/material-components-android/blob/master/catalog/java/io/material/catalog/topappbar/res/layout/cat_topappbar_collapsing_fragment.xml#L30) (updated recently in https://github.com/material-components/material-components-android/commit/d3b7c72446e8b923495383e9c399ad3d1b4d0eab).

johan12345 commented:
> @johan12345 Ah you can set a theme overlay on your AppBarLayout (android:theme="@style/ThemeOverlay.Catalog.AppBarLayout") to fix the text color issue, similar to cat_topappbar_collapsing_fragment.xml (updated recently in d3b7c72).

Okay, then I should better rebase onto master first, right?

> It looks as though there might be a bug in RTL as well? If you enable pseudolocales for the catalog, it causes overlapping text

Indeed, there's something wrong here. I will try to look into that...
I previously only tested the "force RTL layout" option in developer options, which looked fine.

johan12345 commented:
Hmm, I played around with the RTL issue for a couple of hours, but had no success so far. It's really difficult as while the text goes from right to left, the x-positions of things on the display still need to be calculated from the left... Some help would be appreciated!

dsn5ft commented:
> > @johan12345 Ah you can set a theme overlay on your AppBarLayout (android:theme="@style/ThemeOverlay.Catalog.AppBarLayout") to fix the text color issue, similar to cat_topappbar_collapsing_fragment.xml (updated recently in [d3b7c72](https://github.com/material-components/material-components-android/commit/d3b7c72446e8b923495383e9c399ad3d1b4d0eab)).
> 
> Okay, then I should better rebase onto master first, right?

Yep I think the theme overlay was updated after this PR was created.

johan12345 commented:
> > > @johan12345 Ah you can set a theme overlay on your AppBarLayout (android:theme="@style/ThemeOverlay.Catalog.AppBarLayout") to fix the text color issue, similar to cat_topappbar_collapsing_fragment.xml (updated recently in [d3b7c72](https://github.com/material-components/material-components-android/commit/d3b7c72446e8b923495383e9c399ad3d1b4d0eab)).
> > 
> > 
> > Okay, then I should better rebase onto master first, right?
> 
> Yep I think the theme overlay was updated after this PR was created.

Okay, I rebased and applied the ThemeOverlay now 👍

dsn5ft commented:
> > > > @johan12345 Ah you can set a theme overlay on your AppBarLayout (android:theme="@style/ThemeOverlay.Catalog.AppBarLayout") to fix the text color issue, similar to cat_topappbar_collapsing_fragment.xml (updated recently in [d3b7c72](https://github.com/material-components/material-components-android/commit/d3b7c72446e8b923495383e9c399ad3d1b4d0eab)).
> > > 
> > > 
> > > Okay, then I should better rebase onto master first, right?
> > 
> > 
> > Yep I think the theme overlay was updated after this PR was created.
> 
> Okay, I rebased and applied the ThemeOverlay now 👍

Great. Oh and you should be able to remove the `style="?attr/catalogToolbarStyle"` on the `CollapsingToolbarLayout` as well, similar to https://github.com/material-components/material-components-android/commit/d3b7c72446e8b923495383e9c399ad3d1b4d0eab.

johan12345 commented:
I got one part (position of the expanded text) fixed for RTL, but collapsed and cross-section text are still off...

cketcham commented on lib/java/com/google/android/material/internal/CollapsingTextHelper.java:
```diff
@@ -655,8 +657,16 @@ public void draw(Canvas canvas) {
+      if (isRtl) {
+        currentExpandedX = x + textLayout.getLineLeft(0) - expandedFirstLineDrawX * 2 +
+            currentBounds.width() - textLayout.getWidth() * scale;
+        currentCollapsedX = x; //+ currentBounds.width() - collapsedTextWidth * scale;
```
Thanks for working on this!

It looks like `currentCollapsedX` is always just `x` in the end like it was before? Also `collapsedTextWidth` isn't used because it's commented out here. But it does look like this does fix the x position of the expanded text.

johan12345 replied:
Yep, that was an unfinished try to also correct the position of the collapsed text, but it didn't work out this way.

I currently have a new approach where I try to adjust the `calculateBaseOffsets` function to directly calculate the correct separate x positions for the collapsed and expanded text (in LTR text they are the same, but not in RTL because they are aligned on the right side instead of the left) instead of adjusting them afterwards in `draw()`. Probably that approach is much better, but it's also not working correctly yet... should I push it to another branch so that you can take a look?

cketcham replied:
I'm not sure when I'll be able to take a look to try to help identify the problem based on other priorities that we have. But I still think it would be useful to push what you have for now.

johan12345 replied:
Okay, I force-pushed the new approach here and kept the previous one as a backup on our [multiline-collapsingtoolbar-old](https://github.com/material-components/material-components-android/compare/master...opacapp:multiline-collapsingtoolbar-old) branch.

cketcham replied:
Can you rebase the branches on master? We made a change recently which is preventing me from pulling them in. thanks!

johan12345 replied:
Okay, should now both be rebased. I accidentally committed the `pseudoLocalesEnabled true` in the build.gradle, that's probably what caused the conflict. Sorry!

isaklilja commented:
Any updates on this?

johan12345 commented:
@isaklilja As you can see, we are still trying to get it to work correctly with RTL text. I will be on vacation next week, so probably no more updates from me until after that.

shubh10 commented:
@johan12345 @cketcham Any new development on this. Waiting for the multiline support so that we can use in out app.

Thanks.

johan12345 commented:
@shubh10 I have still not found a way to make it work correctly with RTL text. If any of you can help, that would be greatly appreciated. If you do not need RTL support, you can use our current version via JitPack for now, which is what [we are also doing](https://github.com/opacapp/opacclient/commit/4d6e798fe4c45f6c3b245c00903985862ea64d10) currently:

```gradle
// in your dependencies block, replace the material-components-android library with:
implementation 'com.github.opacapp:material-components-android:d290745a71'
// in your repositories block, add:
maven { url "https://jitpack.io" }
```

shubh10 commented:
@johan12345 Thanks for the update. Will use the version provided above in my project.

Also, will investigate about the RTL and try to help with my bit.

ldjcmu commented:
Hi Johan12345,

I just wanted to note that I've added the "Work In Progress" tag to help us better identify PRs that are awaiting changes before merging.

I'd also like to add that we'd really like to add this feature to the library. If it's not apparent how to get the feature to work for RTL, we'd accept the PR if there were default/fallback behavior in that case even if the feature isn't fully functional. I'll also see if we can find time for someone on the team to spend some hours looking at this.

johan12345 commented:
Thanks! I will have a look into whether I can easily implement a fallback behavior (-> show only one line) for RTL text for now.

johan12345 commented:
Hi @ldjcmu , I have pushed a fallback solution for RTL, so that only one line of text is drawn if `isRtl` is `true`.

The code for my previous (second) incomplete attempt at getting it to work correctly with RTL has now moved to here:
https://github.com/material-components/material-components-android/compare/master...opacapp:multiline-collapsingtoolbar-old2

ldjcmu commented:
I'm going to attempt the sync for this now. I'll let you know if anything comes up on our internal tests.

ldjcmu commented:
Hey Johan12345,

Some of our internal tests are showing that the change modifies the bottom of the collapsing title up a good chunk. We'll need to drop this back down to where it was before merging. Do you have an idea as to what could be happening here and if you can keep it the same?

This is the layout we're using for the test. Screenshots attached.

`<?xml version="1.0" encoding="utf-8"?>
<!--
  ~ Copyright (C) 2018 The Android Open Source Project
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->
<androidx.coordinatorlayout.widget.CoordinatorLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:id="@+id/coordinator_layout"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:fitsSystemWindows="false">
  <com.google.android.material.appbar.AppBarLayout
      android:id="@+id/app_bar"
      android:layout_width="match_parent"
      android:layout_height="196dp">

    <com.google.android.material.appbar.CollapsingToolbarLayout
        android:id="@+id/collapsing_toolbar"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        app:collapsedTitleTextAppearance="@style/TextAppearance.Test.CollapsingToolbar.Collapsed"
        app:expandedTitleTextAppearance="@style/TextAppearance.Test.CollapsingToolbar.Expanded"
        app:layout_scrollFlags="scroll|exitUntilCollapsed"
        app:title="@string/toolbar_title">

      <com.google.android.material.appbar.MaterialToolbar
          android:id="@+id/toolbar"
          android:layout_width="match_parent"
          android:layout_height="?attr/actionBarSize"
          app:layout_collapseMode="pin"/>

    </com.google.android.material.appbar.CollapsingToolbarLayout>
  </com.google.android.material.appbar.AppBarLayout>
</androidx.coordinatorlayout.widget.CoordinatorLayout>`

![CollapsingTextMove](https://user-images.githubusercontent.com/42812180/65984899-b95d6b00-e44e-11e9-8b95-fad2dad70106.png)

johan12345 commented:
Hm, it looks like the difference is exactly the ascent of the text. Could it be that it's fixed with this commit? https://github.com/material-components/material-components-android/pull/413/commits/d775c06a3283b56bccb0009762539d0dca8df7d4

ldjcmu commented:
That commit fixed the offset issue, but I have a couple more screenshot diffs. Most of them seem related to the font boldness and alpha or color affecting TextInputLayout. More screendiffs attached.
![diff1](https://user-images.githubusercontent.com/42812180/65999055-9e98ef80-e46a-11e9-8ec6-98453fd761af.png)
![diff2](https://user-images.githubusercontent.com/42812180/65999056-9e98ef80-e46a-11e9-8bb6-18205fa20d4c.png)
![diff3](https://user-images.githubusercontent.com/42812180/65999057-9e98ef80-e46a-11e9-917f-f5d8d94bf183.png)

johan12345 commented:
So, the alpha issue should be fixed with 0b0471f.

However, I can't reprocude the boldness problem. What TextAppearances were used for this test and was the test run in the fully expanded state?

ldjcmu commented:
Thanks for the update. I'll get a re-run of the alpha tests going and see what I can pull for you regarding the boldness.

ldjcmu commented:
expectedViewRule.validateView() takes and compares the screenshots. The test was a bit more involved than I realized with all the styles we were applying.

```
@RunWith(AndroidJUnit4.class)
public class CollapsingToolbarLayoutRenderTest {

  @Rule
  public final ExpectedViewRule<?> expectedViewRule =
      new ExpectedViewRule<>(TestActivity.class);

  @BeforeClass
  public static void beforeClassSetup() {
    TextAppearanceConfig.setShouldLoadFontSynchronously(true);
  }

  @Test
  public void collapsingToolbarLayout_expanded() throws Exception {
    expectedViewRule.layout(R.layout.collapsing_toolbar).validateView();
  }
}
```

Collapsing_toolbar:
```
<?xml version="1.0" encoding="utf-8"?>
<!--
  ~ Copyright (C) 2018 The Android Open Source Project
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->
<androidx.coordinatorlayout.widget.CoordinatorLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:id="@+id/coordinator_layout"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:fitsSystemWindows="false">
  <com.google.android.material.appbar.AppBarLayout
      android:id="@+id/app_bar"
      android:layout_width="match_parent"
      android:layout_height="196dp">

    <com.google.android.material.appbar.CollapsingToolbarLayout
        android:id="@+id/collapsing_toolbar"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        app:collapsedTitleTextAppearance="@style/TextAppearance.Test.CollapsingToolbar.Collapsed"
        app:expandedTitleTextAppearance="@style/TextAppearance.Test.CollapsingToolbar.Expanded"
        app:layout_scrollFlags="scroll|exitUntilCollapsed"
        app:title="@string/toolbar_title">

      <com.google.android.material.appbar.MaterialToolbar
          android:id="@+id/toolbar"
          android:layout_width="match_parent"
          android:layout_height="?attr/actionBarSize"
          app:layout_collapseMode="pin"/>

    </com.google.android.material.appbar.CollapsingToolbarLayout>
  </com.google.android.material.appbar.AppBarLayout>
</androidx.coordinatorlayout.widget.CoordinatorLayout>
```

styles.xml:
```
<resources>

  <style name="TextAppearance.Test.CollapsingToolbar.Expanded" parent="">
    <item name="fontFamily">serif</item>
    <item name="android:fontFamily">serif</item>
    <item name="android:textColor">@android:color/white</item>
    <item name="android:textSize">28sp</item>
    <item name="android:textStyle">italic</item>
  </style>

  <style name="TextAppearance.Test.CollapsingToolbar.Collapsed" parent="">
    <item name="fontFamily">monospace</item>
    <item name="android:fontFamily">monospace</item>
    <item name="android:textSize">20sp</item>
    <item name="android:textColor">@android:color/holo_blue_bright</item>
  </style>
</resources>
```

johan12345 commented:
Hi @ldjcmu, I tried to copy these styles and layout into the catalog sample app to test this. I got this result, which looks the same as on the reference image, as far as I can see:

<img src="https://user-images.githubusercontent.com/5310424/66322865-165f9200-e923-11e9-8afa-4252a34da86a.png" alt="Screenshot_1570459570" width="400">

I don't know why it's turning out different in your tests.

ldjcmu commented:
So it's going to be difficult for us to land this PR with the screendiffs in tests (our team doesn't own all the tests), so I'm thinking of a slightly different approach. Instead of changing the base CollapsingToolbarLayout can we provide this as an opt-in feature? This would leave all the current screenshots unchanged, but still unlock the feature for users who opt-in?

johan12345 commented:
The strange thing is that we haven't touched any of the code that sets the typeface and none of the two TextAppearances makes the text bold. Maybe it has something to do with another test that is run before this one that uses bold text? Does it reproduce the reference image if it is run without our changes?

Hm, we could make it opt-in if that is really necessary, but then the question is how separate it should be:
- A completely new CollapsingToolbarLayout class?
- Two different CollapsingTextHelpers that are switched between if `maxLines > 1` in the CollapsingToolbarLayout
- many additional `if (maxLines > 1) {` clauses within CollapsingTextHelper to switch between the old and new logic

ldjcmu commented:
Hi again, so I had some time to look into what's happening. I did some manual tests (skipping the testing framework for now) and found two potentially related issues that may help with the failing screenshots.

The first - when the title starts with a tab character ('\t') you get this weird double overlap effect.
The second - at the very end of the motion, there's a doubling of the font (looks like an overlap that's a tiny bit off) that looks very similar to the screenshots above.

Videos attached from a nexus 5x running API 28 at 420dpi.



![tabbed-title](https://user-images.githubusercontent.com/42812180/66955138-bc9c4d80-f02f-11e9-997a-ea6a533e25e2.gif)
![double-title](https://user-images.githubusercontent.com/42812180/66955146-be661100-f02f-11e9-9e8b-f2337f547d1d.gif)

johan12345 commented:
Hm, isn't the latter just the switch from regular to bold font when the collapsed textAppearance is set to bold (or medium)?

The first one seems to be a bug (though it's probably quite uncommon to have a title start with a \t...).

matthewdlloydlp commented:
> Hi Johan12345,
> 
> I just wanted to note that I've added the "Work In Progress" tag to help us better identify PRs that are awaiting changes before merging.
> 
> I'd also like to add that we'd really like to add this feature to the library. If it's not apparent how to get the feature to work for RTL, we'd accept the PR if there were default/fallback behavior in that case even if the feature isn't fully functional. I'll also see if we can find time for someone on the team to spend some hours looking at this.

I was just looking through the comments on the thread and found this-- @ldjcmu have the basic requirements to merge this been met, even though it might not be completely perfect? This is functionality we'd love to use in our app!

ldjcmu commented:
> > Hi Johan12345,
> > I just wanted to note that I've added the "Work In Progress" tag to help us better identify PRs that are awaiting changes before merging.
> > I'd also like to add that we'd really like to add this feature to the library. If it's not apparent how to get the feature to work for RTL, we'd accept the PR if there were default/fallback behavior in that case even if the feature isn't fully functional. I'll also see if we can find time for someone on the team to spend some hours looking at this.
> 
> I was just looking through the comments on the thread and found this-- @ldjcmu have the basic requirements to merge this been met, even though it might not be completely perfect? This is functionality we'd love to use in our app!

Unfortunately, we can't merge the code into our internal repo with the failing screenshot tests. It will get reverted in very short order. Right now the fastest solution is to make the change in a way that maintains all existing functionality, but I understand that's frustrating.

isaklilja commented:
> > > Hi Johan12345,
> > > I just wanted to note that I've added the "Work In Progress" tag to help us better identify PRs that are awaiting changes before merging.
> > > I'd also like to add that we'd really like to add this feature to the library. If it's not apparent how to get the feature to work for RTL, we'd accept the PR if there were default/fallback behavior in that case even if the feature isn't fully functional. I'll also see if we can find time for someone on the team to spend some hours looking at this.
> > 
> > 
> > I was just looking through the comments on the thread and found this-- @ldjcmu have the basic requirements to merge this been met, even though it might not be completely perfect? This is functionality we'd love to use in our app!
> 
> Unfortunately, we can't merge the code into our internal repo with the failing screenshot tests. It will get reverted in very short order. Right now the fastest solution is to make the change in a way that maintains all existing functionality, but I understand that's frustrating.

It would be great, if you have time, if you could help @johan12345 by looking into why the tests are producing the results they are. He said something about not touching any textappearances etc, while the test screenshots are showing issues with bold text etc.

AleSpero commented:
+1 for this, thanks for the good work and hope to see it merged soon, this could be really helpful :)

johan12345 commented:
Hi, I have rebased the branch onto the current master and moved our multiline code to a separate `MultilineCollapsingTextHelper` class and also restored the original `CollapsingTextHelper`. In `CollapsingToolbarLayout`, the `multiline` parameter needs to be set to `true` to opt-in for the multiline feature.

Of course, this results in some duplicated code, especially since `MultilineCollapsingTextHelper` is a completely separate class and not a subclass of `CollapsingTextHelper` (which is `final` has many private variables).

But this should mean that the internal tests should work fine as the default behavior will not change compared to the previous version. @ldjcmu, could you check this?

milosmns commented:
I think that some duplication is fine for the rest of us who desperately need this feature built-in. You can create an improvement ticket to clean up the tech debt in the next release, and have a release for us? :)

This PR has gone back and forth for a long time, perhaps its review can be prioritized?

johan12345 commented:
If we would fall back to start, that would mean that if the user specifies only the vertical gravity, this would be overridden. Therefore, I instead mutate the given value to only replace the byte for the horizontal gravity.

ymarian commented:
@johan12345 I understand but you always either check with RELATIVE_HORIZONTAL_GRAVITY_MASK or RELATIVE_VERTICAL_GRAVITY_MASK and you have a default do the same as START so not sure why it's necessary

ymarian commented:
I'm splitting this into more manageable pieces btw, no need to keep updating this pr.

johan12345 commented:
> @johan12345 I understand but you always either check with RELATIVE_HORIZONTAL_GRAVITY_MASK or RELATIVE_VERTICAL_GRAVITY_MASK and you have a default do the same as START so not sure why it's necessary

Hm, you're right. Maybe it was necessary earlier but not in the current state.

ArcherEmiya05 commented:
I was just confuse for the fork library should I use this `com.github.opacapp:material-components-android:d290745a71` or `net.opacapp:multiline-collapsingtoolbar:27.1.1`? Which one is the stable and updated?

johan12345 commented:
`net.opacapp:multiline-collapsingtoolbar:27.1.1` is the latest stable version, but only compatible with Android Design Support Library 27.1.1 (quite old).

`com.github.opacapp:material-components-android:d290745a71` is our new fork that we are trying to get merged into the official material-components-android library. Thus, it is compatible with the newest AndroidX libraries. It also includes some additional fixes that were made after the `net.opacapp:multiline-collapsingtoolbar:27.1.1` release. But, there can be some problems if you use it together with other libraries that depend on material-components-android, because contrary to `net.opacapp:multiline-collapsingtoolbar:27.1.1`, `com.github.opacapp:material-components-android:d290745a71` needs to *replace* the material-components-android library in your `build.gradle`.

ArcherEmiya05 commented:
> material

One more thing can I also add subtitle on the CollapsingToolbar?

ArcherEmiya05 commented:
> @ArcherEmiya05 `net.opacapp:multiline-collapsingtoolbar:27.1.1` is the latest stable version, but only compatible with Android Design Support Library 27.1.1 (quite old).
> 
> `com.github.opacapp:material-components-android:d290745a71` is our new fork that we are trying to get merged into the official material-components-android library with this PR. Thus, it is compatible with the newest AndroidX libraries. It also includes some additional fixes that were made after the `net.opacapp:multiline-collapsingtoolbar:27.1.1` release. But, there can be some problems if you use it together with other libraries that depend on material-components-android, because contrary to `net.opacapp:multiline-collapsingtoolbar:27.1.1`, `com.github.opacapp:material-components-android:d290745a71` needs to _replace_ the material-components-android library in your `build.gradle`.

One more thing, can I also add subtitle on the CollapsingToolbar?

johan12345 commented:
> One more thing, can I also add subtitle on the CollapsingToolbar?

@ArcherEmiya05 That's not something our implementation currently supports. It might be possible to just add a custom TextView to the AppBarLayout for that, but I don't have an example handy.

ArcherEmiya05 commented:
> > One more thing, can I also add subtitle on the CollapsingToolbar?
> 
> @ArcherEmiya05 That's not something our implementation currently supports. It might be possible to just add a custom TextView to the AppBarLayout for that, but I don't have an example handy.

Thanks for the information :)

bazinac commented:
Hi guys, is there chance that these changes might get merged anytime soon? Such a feature (multiline title) will be very handy.

adhirajsinghchauhan commented:
@bazinac There are issues with RTL text, which haven't been fixed yet.

@johan12345 I've managed to add subtitle support as well. What do you think will be better? Finishing this PR up first, and then creating another one for subtitle support? Or shall we do it together?

ymarian commented:
@johan12345 I finally got some time to work on this. Are you interested in continuing from https://github.com/material-components/material-components-android/commit/83e4e47d2225cd281bef195d1baf035117cda4cd ? 

I think there might be a few bugs so I left the API restricted. The current implementation seems to work ok and it's not causing any screenshots changes on our side.
We still need to add the demo and some of the attributes. I wouldn't add all of them to start like spacing and multiplier. Let me know what you think.

Also RTL orientation actually works but an RTL language does not. Let me know if you can take a look and you have other thoughts.

Don't think it's worth implementing jelly bean for now either.

johan12345 commented:
> @johan12345 I've managed to add subtitle support as well. What do you think will be better? Finishing this PR up first, and then creating another one for subtitle support? Or shall we do it together?

@adhirajsinghchauhan Well, in the end, that's something for Google to decide (@ymarian), but I would say this is already complicated enough that subtitles should better be done in a separate PR.

> @johan12345 I finally got some time to work on this. Are you interested in continuing from 83e4e47 ?

Thank you! Yes, I will try to take a look in the next few days.

> I think there might be a few bugs so I left the API restricted. The current implementation seems to work ok and it's not causing any screenshots changes on our side.
> We still need to add the demo and some of the attributes. I wouldn't add all of them to start like spacing and multiplier. Let me know what you think.

Okay, yeah, the demo should be quite straightforward to add.
The `maxLines` attribute is definitely needed. I myself don't need the line spacing customization at the moment, but it was requested (https://github.com/opacapp/multiline-collapsingtoolbar/issues/29) and contributed (https://github.com/opacapp/multiline-collapsingtoolbar/pull/45) to our library, so we should add it at some point.

> Also RTL orientation actually works but an RTL language does not. Let me know if you can take a look and you have other thoughts.

Yes, I think that's what we had seen earlier as well - ["force RTL layout" works, but RTL languages (+ pseudolocales) don't](https://github.com/material-components/material-components-android/pull/413#issuecomment-510208436). RTL is more difficult for this, because in LTR the x positions (which are always calculated from the left) of the collapsed and expanded texts are the same, but not necessarily in RTL. That's why I tried to implement the calculation of the separate positions in https://github.com/opacapp/material-components-android/commit/8f012b3ab577f7b7d5030d3707f5011885dbf1d4, but I haven't got it to work unfortunately.

> Don't think it's worth implementing jelly bean for now either.

Yeah, that's probably fine.

bazinac commented:
I would absolutely love to have subtitle solved as well. At the moment, we are using custom text view in the appBar and its not all that nice. I believe that subtitle would be oftentimes handy feature, due to the nature of the CollapsingToolbarLayout.

Anyways, keep up the good work, I really appreciate your effort and I believe google does so as well :)

johan12345 commented:
So, [on a new branch](https://github.com/material-components/material-components-android/compare/master...opacapp:multiline-collapsingtoolbar-new), I have now added the maxLines attribute and the demo.

ymarian commented:
@johan12345 Thanks a lot, that looks good to me, I can merge that in and close this PR. Any outstanding bugs can be addressed in different PRs or people can file bugs as they start to test.

Sounds good?

johan12345 commented:
That sounds good, thank you! 🎉

katajona commented:
There was one functionality that was lost when integrating this library. I can have multiline text but the width is still limited because of the menu items as you can see here:
<img src="https://user-images.githubusercontent.com/43173721/114382008-1f6e9680-9b8c-11eb-97c7-c05c334f42e5.jpg"  width="300">

In the multilinecollapsingtoolbar library the width was not limited which allowed a lot better design:
<img src="https://user-images.githubusercontent.com/43173721/114381980-167dc500-9b8c-11eb-932f-ace1eb22ba38.jpg"  width="300">
Do you ever plan to put this functionality back? There are a few questions where people are trying to fix this issue(in general for a collapsing toolbar) for example https://stackoverflow.com/questions/27502345/toolbar-title-full-width

johan12345 commented:
@katajona Sorry, I'm no longer working at rami.io, the company who is developing the [opacapp](https://github.com/opacapp/opacclient) project and the multiline-collapsingtoolbar library and the calculations involved in making the multiline animation work smoothly are quite complicated, so I unfortunately don't have time to look into this right now.

I don't remember exactly anymore, but it could be that this functionality was removed either
- (on purpose) to make sure that the result when the multiline functionality is disabled exactly matches the previous version of the library, which was necessary to fulfill Google's internal test suite, or
- (accidentally) in the process of fixing one of the bugs in the original multiline-collapsingtoolbar library, e.g. the missing RTL support
